src/server/scripts/Kalimdor/Firelands/instance_firelands.cpp

```diff
@@ -40,8 +46,59 @@ class instance_firelands : public InstanceMapScript
                         // Cannot directly start attacking here as the creature is not yet on map
                         creature->m_Events.AddEvent(new DelayedAttackStartEvent(creature), creature->m_Events.CalculateTime(500));
                         break;
+                    case NPC_BALEROC:
+                        BalerocGUID = creature->GetGUID();
+                        break;
+                    default:
+                        break;
                 }
             }
+
+            void OnGameObjectCreate(GameObject* go) override
+            {
+                switch(go->GetEntry())
+                {
+                    case GO_BALEROC_FIREWALL:
+                        BalerocDoorGUID = go->GetGUID();
+                        if (GetBossState(DATA_SHANNOX) == DONE || GetBossState(DATA_BALEROC) == DONE)
+                            go->SetGoState(GO_STATE_ACTIVE);
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            bool SetBossState(uint32 type, EncounterState state) override
+            {
+                if (!InstanceScript::SetBossState(type, state))
+                    return false;
+
+                if ((type == DATA_SHANNOX && state == DONE) || (type == DATA_BALEROC && state != IN_PROGRESS))
+                {
+                    if (GameObject* door = instance->GetGameObject(BalerocDoorGUID))
+                        door->SetGoState(GO_STATE_ACTIVE);
+                }
+                else if (type == DATA_BALEROC && state == IN_PROGRESS)
+                    if (GameObject* door = instance->GetGameObject(BalerocDoorGUID))
+                        door->SetGoState(GO_STATE_READY);
```

Comment:
how should it work? it is not exactly clear to me, different states in OnGameObjectCreate and SetBossState